File: backend/src/test/java/org/globe42/web/cities/PostalCityControllerTest.java
package org.globe42.web.cities;

import static org.assertj.core.api.Java6Assertions.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.Arrays;
import java.util.List;

import org.globe42.dao.PostalCityDao;
import org.globe42.domain.PostalCity;
import org.globe42.test.BaseTest;
import org.globe42.web.persons.CityDTO;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;

/**
 * Unit tests for {@link PostalCityController}
 * @author JB Nizet
 */
public class PostalCityControllerTest extends BaseTest {
    @Mock
    private PostalCityDao mockPostalCityDao;

    @Mock
    private PostalCityUploadParser mockUploadParser;

    @InjectMocks
    private PostalCityController controller;

    @Test
    public void shouldSearchByPostalCodeWhenQueryIsNumeric() {
        PostalCity postalCity = new PostalCity("42000", "ST ETIENNE");
        when(mockPostalCityDao.findByPostalCode("420", PostalCityController.LIMIT)).thenReturn(
            Arrays.asList(postalCity)
        );

        List<CityDTO> result = controller.search("420");

        assertThat(result).hasSize(1);
        assertThat(result.get(0).getCode()).isEqualTo(postalCity.getPostalCode());
        assertThat(result.get(0).getCity()).isEqualTo(postalCity.getCity());
    }

    @Test
    public void shouldSearchByCityWhenQueryIsNotNumeric() {
        PostalCity postalCity = new PostalCity("42000", "ST ETIENNE");
        when(mockPostalCityDao.findByCity("ST ET", PostalCityController.LIMIT)).thenReturn(
            Arrays.asList(postalCity)
        );

        List<CityDTO> result = controller.search("ST ET");

        assertThat(result).hasSize(1);
        assertThat(result.get(0).getCode()).isEqualTo(postalCity.getPostalCode());
        assertThat(result.get(0).getCity()).isEqualTo(postalCity.getCity());
    }

    @Test
    public void shouldUpload() {
        byte[] body = "fake".getBytes();
        List<PostalCity> parsedCities = Arrays.asList(new PostalCity("42000", "ST ETIENNE"));
        when(mockUploadParser.parse(body)).thenReturn(parsedCities);

        controller.upload(body);

        verify(mockPostalCityDao).saveAllEfficiently(parsedCities);
    }
}

File: backend/src/test/kotlin/org/globe42/web/cities/PostalCityControllerTest.kt
package org.globe42.web.cities

import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.globe42.dao.PostalCityDao
import org.globe42.domain.PostalCity
import org.junit.jupiter.api.Test
import java.util.*

/**
 * Unit tests for [PostalCityController]
 * @author JB Nizet
 */
class PostalCityControllerTest {
    private val mockPostalCityDao = mockk<PostalCityDao>(relaxUnitFun = true)

    private val mockUploadParser = mockk<PostalCityUploadParser>()

    private val controller = PostalCityController(mockPostalCityDao, mockUploadParser)

    @Test
    fun `should search by postal code when query is numeric`() {
        val postalCity = PostalCity("42000", "ST ETIENNE")
        every { mockPostalCityDao.findByPostalCode("420", LIMIT) } returns listOf(postalCity)

        val result = controller.search("420")

        assertThat(result).hasSize(1)
        assertThat(result[0].code).isEqualTo(postalCity.postalCode)
        assertThat(result[0].city).isEqualTo(postalCity.city)
    }

    @Test
    fun `should search by city when query is not numeric`() {
        val postalCity = PostalCity("42000", "ST ETIENNE")
        every { mockPostalCityDao.findByCity("ST ET", LIMIT) } returns listOf(postalCity)

        val result = controller.search("ST ET")

        assertThat(result).hasSize(1)
        assertThat(result[0].code).isEqualTo(postalCity.postalCode)
        assertThat(result[0].city).isEqualTo(postalCity.city)
    }

    @Test
    fun `should upload`() {
        val body = "fake".toByteArray()
        val parsedCities = Arrays.asList(PostalCity("42000", "ST ETIENNE"))
        every { mockUploadParser.parse(body) } returns parsedCities

        controller.upload(body)

        verify { mockPostalCityDao.saveAllEfficiently(parsedCities) }
    }
}
